docs/design/proposals/resource-reference/resource-reference.md

```diff
@@ -0,0 +1,526 @@
+
+	opMap := m.SDKAPI.GetOperationMap(m.cfg)
+
+	createOps := (*opMap)[OpTypeCreate]
+	readOneOps := (*opMap)[OpTypeGet]
+	readManyOps := (*opMap)[OpTypeList]
+	updateOps := (*opMap)[OpTypeUpdate]
+	deleteOps := (*opMap)[OpTypeDelete]
+	getAttributesOps := (*opMap)[OpTypeGetAttributes]
+	setAttributesOps := (*opMap)[OpTypeSetAttributes]
+
+	for crdName, createOp := range createOps {
+		if m.cfg.IsIgnoredResource(crdName) {
+			continue
+		}
+		crdNames := names.New(crdName)
+...
+		// Now any additional Spec fields that are required from other API
+		// operations.
+		for targetFieldName, fieldConfig := range m.cfg.ResourceFields(crdName) {
+			if fieldConfig.IsReadOnly {
+				// It's a Status field...
+				continue
+			}
+
++			if fieldConfig.Reference != nil {
++				// Add the spec field here
++				referenceFieldNames := names.New(targetFieldName)
++				rf := NewReferenceField(crd, referenceFieldNames, fieldConfig)
++				crd.SpecFields[referenceFieldNames.Original] = rf
++				crd.Fields[referenceFieldNames.Camel] = rf
++			}
+
+...
+	m.crds = crds
+	return crds, nil
+}
+```
+
+```go
+//field.go
+func NewReferenceField(
+	crd *CRD,
+	fieldNames names.Names,
+	cfg *ackgenconfig.FieldConfig,
+) *Field {
+	gt := "*ackv1alpha1.AWSResourceReference"
+	gtp := "*ackv1alpha1.AWSResourceReference"
+	if cfg.Reference.IsList {
+		gt = "[]" + gt
+		gtp = "[]" + gtp
+	}
+	return &Field{
+		CRD:               crd,
+		Names:             fieldNames,
+		Path:              fieldNames.Original,
+		ShapeRef:          nil,
+		GoType:            gt,
+		GoTypeElem:        "AWSResourceReference",
+		GoTypeWithPkgName: gtp,
+		FieldConfig:       cfg,
+	}
+}
+```
+
+#### Runtime Changes
+
+```go
+// aws_resource_manager.go
+
+type AWSResourceManager interface {
+        ...
++       // DereferenceResource finds if there are any AWSResourceReference fields
++       // present in the AWSResource passed in the parameters and attempts to resolve
++       // those reference fields into target field.
++       // It returns an AWSResource with resolved references, boolean representing
++       // whether any reference fields were present and an error if the passed
++       // AWSResource's reference fields cannot be resolved
++       DereferenceResource(context.Context, client.Reader, AWSResource) (AWSResource, bool, error)
+}
+```
+
+```go
+// reconciler.go
+
+// reconcile either cleans up a deleted resource or ensures that the supplied
+// AWSResource's backing API resource matches the supplied desired state.
+//
+// It returns a copy of the resource that represents the latest observed state.
+func (r *resourceReconciler) reconcile(
+ctx context.Context,
+rm acktypes.AWSResourceManager,
+res acktypes.AWSResource,
+) (acktypes.AWSResource, error) {
+    if res.IsBeingDeleted() {
++       // Resolve references before deleting the resource.
++       // Ignore any errors while resolving the references
++    	  res, _, _ = rm.DereferenceResource(ctx, r.apiReader, res)
```

Comment:
I think we should ignore deletion in this proposal as it's a whole other can of worms to get into.

Reply:
Yep, but we do need the referenced field in order to construct `DeleteInput`

Reply:
Should we be storing the dereferenced value in some way after we create the resource? If the referenced resource changes after we create the resource that references it, we lose context about what the original Create call contained?

Reply:
> Should we be storing the dereferenced value in some way after we create the resource?

I don't think so. I did not persist the dereferenced value back into spec either.

> If the referenced resource changes after we create the resource that references it, we lose context about what the original Create call contained?

Currently, we are only using referenced resources to find identifiers which do not change when referenced resource is updated. This was the basis on which I skipped implementing watchers for referenced resource.

So, ATM i don't think we need to preserve anything from dereferenced field. It is also NOT a one way door decision.